video/out/vo_tct.c

```diff
@@ -262,17 +248,19 @@ static void flip_page(struct vo *vo)
     if (vo->dwidth != width || vo->dheight != height)
         reconfig(vo, vo->params);
 
+    p->frame_buf.len = 0;
     if (p->opts.algo == ALGO_PLAIN) {
-        write_plain(
+        write_plain(&p->frame_buf,
             vo->dwidth, vo->dheight, p->swidth, p->sheight,
             p->frame->planes[0], p->frame->stride[0],
             p->opts.term256, p->lut);
     } else {
-        write_half_blocks(
+        write_half_blocks(&p->frame_buf,
             vo->dwidth, vo->dheight, p->swidth, p->sheight,
             p->frame->planes[0], p->frame->stride[0],
             p->opts.term256, p->lut);
     }
+    printf("%.*s\n", BSTR_P(p->frame_buf));
```

Comment:
Why printf here instead of fwrite? It may be necessary on windows to translate from UTF-8 to the codepage, but if so, shouldn't we have an `#ifdef _WIN32` or something as above?

Reply:
It is done once per frame, I don't think there is measurable difference.

Reply:
The `printf` variant still needs to do `strlen` (or something to that effect), because it won't write past the end of the string. The precision value doesn't imply ignoring the terminating null if it arrives before the precision is reached.

If you have the length and it doesn't need any of the things which printf does (like interpret format specifiers, or align, etc), then just [f]write it directly.

Reply:
> The printf variant still needs to do strlen (or something to that effect), because it won't write past the end of the string. The precision value doesn't imply ignoring the terminating null if it arrives before the precision is reached.

I'm pretty sure lying to printf about string length is basically UB and likely it just blow up and ignore the thing. But sure, no need to call it at all.

Reply:
> I'm pretty sure lying to printf about string length

You can't lie about the string length because the precision value is not meant to specify the string length. With `%s` and `%S`, it specifies the maximum number of bytes to output (in cases where the string length is more than that).

Reply:
fixed